test/AltRsmiTests.cpp

```diff
@@ -0,0 +1,932 @@
+
+// ============================================================================
+// Test utilities for manipulating alt_rsmi.cc internal state
+// ============================================================================
+namespace AltRsmiTestUtils {
+
+// Set the KFD nodes path for testing
+// This redirects file reads to test directories
+void SetNodesPath(const char* path) {
+    kKFDNodesPathRoot = path;
+}
+
+// Reset ARSMI internal state between tests
+// This ensures test isolation
+void ResetState() {
+    ARSMI_num_devices = -1;
+    ARSMI_orderedNodes.clear();
+    ARSMI_orderedLinks.clear();
+}
+
+// Get current number of devices (for verification)
+int GetNumDevices() {
+    return ARSMI_num_devices;
+}
+
+} // namespace AltRsmiTestUtils
+
+// Test paths for creating mock KFD filesystem
+static const char* kTestKFDBasePath = "/tmp/test_kfd_arsmi";
+// IMPORTANT: This must match the default value in alt_rsmi.cc
+// Always restore this path after tests to avoid affecting other test suites
+static const char* kSystemKFDPath = "/sys/class/kfd/kfd/topology/nodes";
+static const char* kTestKFDPath = "/tmp/test_kfd_arsmi/topology/nodes";
+
+namespace RcclUnitTesting {
+
+// Global test environment to ensure path restoration even if tests fail
+class AltRsmiTestEnvironment : public ::testing::Environment {
+public:
+  ~AltRsmiTestEnvironment() override = default;
+
+  // Called before any tests run
+  void SetUp() override {
+    // Save the original path (should be system default)
+    originalPath = kSystemKFDPath;
+  }
+
+  // Called after all tests finish
+  void TearDown() override {
+    // Ensure path is restored to system default
+    // This catches any test that failed to restore in its TearDown()
+    AltRsmiTestUtils::SetNodesPath(kSystemKFDPath);
+    AltRsmiTestUtils::ResetState();
+  }
+
+private:
+  const char* originalPath;
+};
+
+// Register the global test environment
+// This ensures cleanup happens even if individual tests fail
+static ::testing::Environment* const altRsmiEnv =
+    ::testing::AddGlobalTestEnvironment(new AltRsmiTestEnvironment());
+
+class AltRsmiTest : public ::testing::Test {
+
+protected:
+  // Helper function to create directories recursively
+  int createDirectory(const std::string &path) {
+    size_t pos = 0;
+    std::string currentPath;
+
+    // Iterate through each component of the path
+    while ((pos = path.find('/', pos)) != std::string::npos) {
+      currentPath = path.substr(0, pos++);
+      if (!currentPath.empty() && mkdir(currentPath.c_str(), 0700) == -1 &&
+          errno != EEXIST) {
+        return -1; // Return error if directory creation fails
+      }
+    }
+
+    // Create the final directory
+    if (mkdir(path.c_str(), 0700) == -1 && errno != EEXIST) {
+      return -1; // Return error if directory creation fails
+    }
+
+    return 0; // Success
+  }
+
+  // Helper function to remove a directory recursively
+  int removeDirectory(const std::string &path) {
```

Comment:
`createDirectory` and `removeDirectory` can be `static`
```suggestion
  static int createDirectory(const std::string &path) {
    size_t pos = 0;
    std::string currentPath;

    // Iterate through each component of the path
    while ((pos = path.find('/', pos)) != std::string::npos) {
      currentPath = path.substr(0, pos++);
      if (!currentPath.empty() && mkdir(currentPath.c_str(), 0700) == -1 &&
          errno != EEXIST) {
        return -1; // Return error if directory creation fails
      }
    }

    // Create the final directory
    if (mkdir(path.c_str(), 0700) == -1 && errno != EEXIST) {
      return -1; // Return error if directory creation fails
    }

    return 0; // Success
  }

  // Helper function to remove a directory recursively
  static int removeDirectory(const std::string &path) {
```